Email 2FA is only fired after retry

Hi, I tried enabling Email 2FA (with sqlite), and it works well, but there is a strange behaviour.
I never receive the email after a login try, every time a need to ask for an other mail to get the code, which then works fine...

By the way, to keep consistency with other parameters : a way to enable email 2FA with environment variables would be nice.

Best regards, and thanks for that amazing work.

Hello,

Thanks for your response:

The server log obtained after a login without first mail sent and with second mail sent after asking is :
```
[2019-10-02 15:34:14][rocket::rocket][INFO] POST /api/accounts/prelogin application/json; charset=utf-8:
[2019-10-02 15:34:14][_][INFO] Matched: POST /api/accounts/prelogin (prelogin)
[2019-10-02 15:34:14][_][INFO] Outcome: Success
[2019-10-02 15:34:14][_][INFO] Response succeeded.
[2019-10-02 15:34:14][rocket::rocket][INFO] POST /identity/connect/token application/x-www-form-urlencoded; charset=utf-8:
[2019-10-02 15:34:14][_][INFO] Matched: POST /identity/connect/token (login)
[2019-10-02 15:34:14][bitwarden_rs::error][ERROR] JsonError. {"TwoFactorProviders":[1],"TwoFactorProviders2":{"1":{"Email":"ni*****@*****.fr"}},"error":"invalid_grant","error_description":"Two factor required."}
[2019-10-02 15:34:14][_][INFO] Outcome: Success
[2019-10-02 15:34:14][_][INFO] Response succeeded.
[2019-10-02 15:34:17][rocket::rocket][INFO] GET /alive:
[2019-10-02 15:34:17][_][INFO] Matched: GET /alive (alive)
[2019-10-02 15:34:17][_][INFO] Outcome: Success
[2019-10-02 15:34:17][_][INFO] Response succeeded.
[2019-10-02 15:34:46][rocket::rocket][INFO] POST /api/two-factor/send-email-login application/json; charset=utf-8:
[2019-10-02 15:34:46][_][INFO] Matched: POST /api/two-factor/send-email-login (send_email_login)
[2019-10-02 15:34:47][lettre::smtp][INFO] connection established to 51.15.206.251:465
[2019-10-02 15:34:47][lettre::smtp][INFO] 9ece511a-99e8-4c22-8de9-0dae45be29af: from=<system@******.fr>
[2019-10-02 15:34:47][lettre::smtp][INFO] 9ece511a-99e8-4c22-8de9-0dae45be29af: to=<ni*****@*********.fr>
[2019-10-02 15:34:47][rocket::rocket][INFO] GET /alive:
[2019-10-02 15:34:47][_][INFO] Matched: GET /alive (alive)
[2019-10-02 15:34:47][_][INFO] Outcome: Success
[2019-10-02 15:34:47][_][INFO] Response succeeded.
[2019-10-02 15:34:55][lettre::smtp][INFO] 9ece511a-99e8-4c22-8de9-0dae45be29af: conn_use=1, status=sent (2.0.0 Ok: queued as 4F0F4A01E6)
[2019-10-02 15:34:55][_][INFO] Outcome: Success
[2019-10-02 15:34:55][_][INFO] Response succeeded.
[2019-10-02 15:35:00][rocket::rocket][INFO] POST /notifications/hub/negotiate text/plain; charset=UTF-8:
[2019-10-02 15:35:00][_][INFO] Matched: POST /notifications/hub/negotiate (negotiate)
[2019-10-02 15:35:00][_][INFO] Outcome: Success
[2019-10-02 15:35:00][_][INFO] Response succeeded.
```
There is actually an error on line 7...

Bitwarden is started with sqlite and the following code :
```
docker run -d --name bitwarden \
    -v /etc/localtime:/etc/localtime:ro \
    -v /etc/timezone:/etc/timezone:ro \
    -v $DATA_FOLDER/:/data/ \
    --network=mailcowdockerized_mailcow-network \
    --restart unless-stopped \
    -e SIGNUPS_ALLOWED=false \
    -e ADMIN_TOKEN=$BW_ADMIN_TOKEN \
    -e SMTP_HOST=****.*****.fr \
    -e SMTP_FROM=system@*******.fr \
    -e SMTP_PORT=465 \
    -e SMTP_SSL=true \
    -e SMTP_EXPLICIT_TLS=true \
    -e SMTP_USERNAME=system@**********.fr \
    -e SMTP_PASSWORD=$PASSWORD \
    -e DOMAIN=https://*******.******.fr \
    -e _ENABLE_EMAIL_2FA=true \
    -e INPUT_EMAIL_TOKEN_SIZE=6 \
    -e INPUT_EMAIL_EXPIRATION_TIME=600 \
    -e INPUT_EMAIL_ATTEMPTS_LIMIT=3 \
    bitwardenrs/server:latest
```

```
[2019-10-15 06:36:57][bitwarden_rs::error][ERROR] JsonError.
{
    "TwoFactorProviders": [
        0,
        1,
        2,
        4
    ],
    "TwoFactorProviders2": {
        "0": null,
        "1": {
            "Email": "some@email.test"
        },
        "2": {
            "Host": "somehost.duosecurity.com",
            "Signature": "signature"
        },
        "4": {
            "Challenges": "[{\"appId\":\"https://bitwarden.somehost.test/app-id.json\",\"challenge\":\"challenge\",\"keyHandle\":\"keyhandle\",\"version\":\"U2F_V2\"}]"
        }
    },
    "error": "invalid_grant",
    "error_description": "Two factor required."
}
```

The error seems normal. It's a message back to the client to say that logging in needs two factor authentication.

I can reproduce the error when I only have the email 2FA enabled.
Working on a fix.

--- a/src/api/core/two_factor/email.rs
+++ b/src/api/core/two_factor/email.rs
@@ -55,17 +55,18 @@ fn send_email_login(data: JsonUpcase<SendEmailLoginData>, conn: DbConn) -> Empty
         err!("Email 2FA is disabled")
     }
 
-    let type_ = TwoFactorType::Email as i32;
-    let mut twofactor = TwoFactor::find_by_user_and_type(&user.uuid, type_, &conn)?;
-
-    prepare_send_token(&mut twofactor, &conn)?;
+    send_token(&user.uuid, &conn)?;
 
     Ok(())
 }
 
 /// Generate the token, save the data for later verification and send email to user
-pub fn prepare_send_token(twofactor: &mut TwoFactor, conn: &DbConn) -> EmptyResult {
+pub fn send_token(user_uuid: &str, conn: &DbConn) -> EmptyResult {
+    let type_ = TwoFactorType::Email as i32;
+    let mut twofactor = TwoFactor::find_by_user_and_type(user_uuid, type_, &conn)?;
+
     let generated_token = generate_token(CONFIG.email_token_size())?;
+
     let mut twofactor_data = EmailTokenData::from_json(&twofactor.data)?;
     twofactor_data.set_token(generated_token);
     twofactor.data = twofactor_data.to_json();

--- a/src/api/core/two_factor/email.rs
+++ b/src/api/core/two_factor/email.rs
@@ -58,6 +58,13 @@ fn send_email_login(data: JsonUpcase<SendEmailLoginData>, conn: DbConn) -> Empty
     let type_ = TwoFactorType::Email as i32;
     let mut twofactor = TwoFactor::find_by_user_and_type(&user.uuid, type_, &conn)?;
 
+    prepare_send_token(&mut twofactor, &conn)?;
+
+    Ok(())
+}
+
+/// Generate the token, save the data for later verification and send email to user
+pub fn prepare_send_token(twofactor: &mut TwoFactor, conn: &DbConn) -> EmptyResult {
     let generated_token = generate_token(CONFIG.email_token_size())?;
     let mut twofactor_data = EmailTokenData::from_json(&twofactor.data)?;
     twofactor_data.set_token(generated_token);

--- a/src/api/identity.rs
+++ b/src/api/identity.rs
@@ -293,13 +293,19 @@ fn _json_err_twofactor(providers: &[i32], user_uuid: &str, conn: &DbConn) -> Api
             }
 
             Some(tf_type @ TwoFactorType::Email) => {
-                let twofactor = match TwoFactor::find_by_user_and_type(user_uuid, tf_type as i32, &conn) {
+                use crate::api::core::two_factor as _tf;
+
+                let mut twofactor = match TwoFactor::find_by_user_and_type(user_uuid, tf_type as i32, &conn) {
                     Some(tf) => tf,
                     None => err!("No twofactor email registered"),
                 };
 
-                let email_data = EmailTokenData::from_json(&twofactor.data)?;
+                // Send email immediately if email is the only 2FA option
+                if providers.len() == 1 {
+                    _tf::email::prepare_send_token(&mut twofactor, &conn)?
+                }
 
+                let email_data = EmailTokenData::from_json(&twofactor.data)?;
                 result["TwoFactorProviders2"][provider.to_string()] = json!({
                     "Email": email::obscure_email(&email_data.email),
                 })

--- a/src/api/identity.rs
+++ b/src/api/identity.rs
@@ -295,14 +295,14 @@ fn _json_err_twofactor(providers: &[i32], user_uuid: &str, conn: &DbConn) -> Api
             Some(tf_type @ TwoFactorType::Email) => {
                 use crate::api::core::two_factor as _tf;
 
-                let mut twofactor = match TwoFactor::find_by_user_and_type(user_uuid, tf_type as i32, &conn) {
+                let twofactor = match TwoFactor::find_by_user_and_type(user_uuid, tf_type as i32, &conn) {
                     Some(tf) => tf,
                     None => err!("No twofactor email registered"),
                 };
 
                 // Send email immediately if email is the only 2FA option
                 if providers.len() == 1 {
-                    _tf::email::prepare_send_token(&mut twofactor, &conn)?
+                    _tf::email::send_token(&user_uuid, &conn)?
                 }
 
                 let email_data = EmailTokenData::from_json(&twofactor.data)?;

